### DISCUSSION TITLE
Introducing Viewing Plots in Editors!

### DISCUSSION THREAD

**Discussion by timtmok** [posit-dev/positron#5073] [Announcements]
You may have noticed that since [2024.09.0-27](https://github.com/posit-dev/positron/releases/tag/2024.09.0-27), there is a new option to open a plot in an editor tab for viewing! Let’s see how it can be enabled and what to expect when using this feature.

### Enable the Feature
By default, the ability to open a plot in an editor is disabled for now. So if you want to try it out, go to Positron’s preferences and search for ‘plots’. The experimental feature will appear and all you need to do is enable it.

![feature_enable](https://github.com/user-attachments/assets/1f1288b0-5276-4897-b5a7-7cd499fa36b6)

If you like editing the preferences JSON file, you can add `"application.experimental.positronPlotsInEditorTab": true` to the file.

### Opening a Plot in an Editor
Once the feature has been enabled, a new button will appear in the Plots view action bar. It is visible when a plot can be opened in an editor. Render a plot and see if the button is available! It will work with both Python and R plots. Interactive plots that render as HTML aren’t supported at this time.

![opening_plot](https://github.com/user-attachments/assets/114d96d3-c9a3-4a38-8558-18b3981bee30)

This is what it looks like when a plot is in an editor. Currently, Positron assigns an id to each plot and that’s what you see in the editor tab title. The plot size and shape does use the same setting that is in the Plots view. So changing the size and shape from the view will also affect any plots in an editor. Right now it’s limited with what you can do with the plot but that will change soon! For now, enjoy viewing the plot in a new light.

![plot_in_editor](https://github.com/user-attachments/assets/a8084491-64a5-4882-b604-c249697dc999)

Here’s what you can do with the plot as of 2024.10.0:
* Split the editor tab to view the plot and other editors side-by-side. Right-click on the editor tab and click on Split Up/Down/Left/Right or drag the editor tab to a new location.

https://github.com/user-attachments/assets/1aef2dc1-f3c6-4859-a539-f18c2338271a

* Float the editor tab in its own window. Right-click on the editor tab and click on Move into New Window. You can also copy it to a new window or split it but it shares the same plot instance so it may behave oddly.

![float_plot](https://github.com/user-attachments/assets/69e42f6f-45cf-4d71-be26-9938f979a883)

* Copy the plot to clipboard. There is an action on the editor tab’s action bar located at the right side of the editor title area. This copies what you see in the editor to the clipboard.

Also, the copy plot to clipboard action had an overhaul. If you invoke it from the command palette (Cmd/Ctrl + Shift + P) while one or more plots are in an editor, you can select which plot you want to copy. The quick pick only appears when there is more than one plot available to copy. Otherwise, it copies the currently selected plot.

Feedback and What’s Next
As this feature gets more exposure, and before it moves out of experimental status, it will get more capabilities. This is what is planned for the future:
* Plot sizing for each editor - The setting is currently tied to the Plots View but the plan is to allow each editor to have its own size. [#4358](https://github.com/posit-dev/positron/issues/4358)
* Zoom - You will be able to scale the plot in the editor. [#4359](https://github.com/posit-dev/positron/issues/4359)
* Save - You will be able to save exactly what you see in the editor in the currently supported formats. [#4361](https://github.com/posit-dev/positron/issues/4361)

If you like what you see in this feature and want to give some feedback, please leave a comment here!

**Comment by LR1933** [comment]
Thank you for your great work! It is really beautiful.
I have a small suggestion. Can you add a button in the Plots view action bar for float the editor tab to a new window?
Currently it takes two presses to get the new window to pop up, if could, a separate quick button would be helpful. 
Thank you very much.

**Reply by timtmok** [reply]
That's a really good suggestion and we have thought about that before. We actually have an icon in mind already. We hope to get this in the Plots view soon!

I've created https://github.com/posit-dev/positron/issues/5446 to track this.

**Reply by timtmok** [reply]
Changes have merged for #5446. This turns the button into a menu button where there are choices for opening in an editor, new window, or to the side. Once the selection has been made, it will be saved so the menu button can be clicked to invoke the preferred action.

**Comment by vinaychuri** [comment]
This is excellent, Thank you!!

It would be interesting to see what users prefer on the "zoom plot" button click ? 
Plot in new editor tab or Plot in new window popup

My vote is plot in new window pop

**Reply by timtmok** [reply]
That is an interesting thought and maybe a preference would fulfill both kinds of users. Thank you for your feedback!

**Comment by jdonland** [comment]
This is excellent, and resolves my only real pain point in transitioning between RStudio and Positron.

**Comment by padpadpadpad** [comment]
Would it be possible for plots in a new editor to update when a new plot appears in the plot pane?

**Reply by juliasilge** [reply]
Can you say a little more about what you mean here? You were thinking about _replacing_ the plot in the editor tab when you make a new plot? So for example, with R:

- `plot(1:10)`
- Open to an editor tab from the Plots pane
- `plot(11:20)`
- The plot shows up in the Plots pane, and then also would... replace the first plot in the existing editor tab?

Is that what you were thinking? Maybe say a little more about the end goal you have; perhaps the Plots pane is too small for you and you always want the bigger space?

**Reply by padpadpadpad** [reply]
Hi @juliasilge. So sometimes in RStudio I have Plot Zoom on a different screen or separate from the plots pane, and it updates as I iterate through new plots. However, in Positron once you open the plot in a new window it is only a single plot and does not autoupdate as new plots occur, unless I am missing something. I used to very much like that behaviour of RStudio.

**Reply by jmcphers** [reply]
@padpadpadpad I believe we're tracking this feature request in https://github.com/posit-dev/positron/issues/7054. Comment/upvote there if you're interested in this!

**Comment by manwarA** [comment]
Thank you for this excellent feature and IDE. However, when I create plot, some plots appear in the plot panel, while for others, I am having this issue:

```
"FigureCanvasAgg is non-interactive, and thus cannot be shown".
```

Its been a while that I am trying to figure out what's wrong with the setting or way I am plotting, however, I could not find reasonable answer from google, SO or reddit. Even the simplest thing like very basic plot is showing the error

```
>>> plt.plot([1,2,3,4,5])
[<matplotlib.lines.Line2D at 0x2213533c6b0>]
>>> plt.show()
<positron-console-cell-217>:1: UserWarning:

FigureCanvasAgg is non-interactive, and thus cannot be shown
```

After several failed attempts, I am seeking help so that I can rectify this issue. I am sorry, if it has been solved, but I could not find anything. My positron is latest. 

Positron Version: 2025.11.0 (user setup) build 234
Code - OSS Version: 1.105.0
Commit: 807acdfca108da6b9b2d96452fe047238da1bd9f
Date: 2025-10-31T23:21:13.588Z
Electron: 37.6.0
Chromium: 138.0.7204.251
Node.js: 22.19.0
V8: 13.8.258.32-electron.0
OS: Windows_NT x64 10.0.26200

**Reply by juliasilge** [reply]
Can you open a [new discussion](https://github.com/posit-dev/positron/discussions) about the problem you are observing, and include code you are using to generate the plot plus info on the Python environment you are using? We will see if we can get to the bottom of it!

**Reply by manwarA** [reply]
Hi, accidently restarted the computer, and now it is working fine. However, this is not the first time that I am facing this issue, I will try to re-create the problem (if there is any), so that I should be fixed. If I come across this again, I will open an issue. Thank you for your time.
